### ISSUE TITLE
Support AssumeRoleWithWebIdentity for AWS provider

### ISSUE THREAD

**Issue by vladiceanu** [apache/airflow#12035]
**Description**
Support of [AssumeRoleWithWebIdentity](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sts.html#STS.Client.assume_role_with_web_identity) for AWS provider when running Airflow workers in EKS. 

**Use case / motivation**
This feature will allow us to use [IRSA](https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/) with Airflow Pods running on EKS and x-account _assumeRole_.

A connection type aws with empty username & password and the following extra parameters
```json
{
"role_arn": "<role_arn>",
"region_name": "<region>",
"aws_session_token": "file://$AWS_WEB_IDENTITY_TOKEN_FILE"
}
```
will retrieve temporary credentials using method `sts-assume-role-with-web-identity` (see [boto3 documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sts.html#STS.Client.assume_role_with_web_identity) of this method)

**Comment by mik-laj** [comment]
Is it related? https://github.com/apache/airflow/issues/9873

**Comment by vladiceanu** [comment]
> Is it related? #9873 If so, I will start working on it this week or next.

not quite sure if the issue you've mentioned includes this one, but seems to include similar functionality on AWS provider. Please note that in this request we're aiming to be able not only to use the simple method `assumeRole`, but also `assumeRoleWithWebIdentity` when Airflow runs on EKS.

**Comment by mik-laj** [comment]
@vladiceanu This looks like a related problem to me, but in my case, but I will want to exchange the token obtained from GKE Workfload Identity with an AWS token. I will not be able to test it with EKS, but I will try to prepare the code in such a way that it is possible to add another token provider.

Can I assign you to this ticket so that you can write the missing EKS integration?

**Comment by vladiceanu** [comment]
> @vladiceanu This looks like a related problem to me, but in my case, but I will want to exchange the token obtained from GKE Workfload Identity with an AWS token. I will not be able to test it with EKS, but I will try to prepare the code in such a way that it is possible to add another token provider.
> 
> Can I assign you to this ticket so that you can write the missing EKS integration?

Yes, please. Thank you

**Comment by mik-laj** [comment]
My PR has been merged. Do you need help to add integrations with EKS?

**Comment by vladiceanu** [comment]
> My PR has been merged. Do you need help to add integrations with EKS?

I'm going to work on it by the end of this week. Thank you.

**Comment by vladiceanu** [comment]
sorry folks, I've got a lot on my plate right now and couldn't work on this. I'll unnasign in case if anyone wants to work on it and come back to it by the end of December if nobody picks it up. Cheers

**Comment by hammad13060** [comment]
@mik-laj Is anyone working on it? I'd like to give it a shot.

**Comment by mik-laj** [comment]
@vladiceanu @hammad13060  Isn't it enough to configure an empty AWS connection to use [the boto3 credential search flow](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#configuring-credentials) and set the AWS_WEB_IDENTITY_TOKEN_FILE environment variable?

Reference:
https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/

**Comment by kloty** [comment]
> @vladiceanu @hammad13060 Isn't it enough to configure an empty AWS connection to use [the boto3 credential search flow](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#configuring-credentials) and set the `AWS_WEB_IDENTITY_TOKEN_FILE` environment variable?
> 
> Reference:
> https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/
> 
> CC: @besson, @arobinski, @kokojumbo, @SadatAnwar, @Swalloow, and @giannisbetas

For me, this method worked with the correct settings. I'll leave my case for the next person. 

I installed Airflow 2.1.2 with helm and attached the required role-arn to the airflow service account as an annotation. The role Includes s3 read write policy for logging and other aws resource related policies. 
I also checked IRSA related Environment Variable (`AWS_ROLE_ARN`, `AWS_WEB_IDENTITY_TOKEN_FILE`) were automatically attached to the Airflow-related pods well.

According to `airflow/providers/amazon/aws/hooks/base_aws.py`, if `role-arn` does not specify in the aws connection's extra parameter, a base session declared using an empty parameter will be used.
So I declared AWS Connection with name aws_default without any additional parameters, and I confirmed that everything worked fined in my project.

However, there was some confusion when reading and working on documentation.
:https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/connections/aws.html#amazon-web-services-connection
Because there are only the GCP related contents are written.

I don't know if above should be written down on documentation, but if that part is resolved, I think this issue can be closed.
